lib/src/embedded/isolate_dispatcher.dart

```diff
@@ -102,84 +106,67 @@ class IsolateDispatcher {
         .transform(const ExplicitCloseTransformer());
-    channel.stream.listen(null,
-        onError: (Object error, StackTrace stackTrace) =>
-            _handleError(error, stackTrace),
-        onDone: _channel.sink.close);
-    return _activate(channel, compilationId, resource);
-  }
-
-  /// Activates [isolate] for a new individual compilation.
-  ///
-  /// This pipes all the outputs from the given isolate through to [_channel].
-  /// The [resource] is released once the isolate is no longer active.
-  StreamSink<Uint8List> _activate(StreamChannel<_InitialMessage> isolate,
-      int compilationId, PoolResource resource) {
-    _inactiveIsolates.remove(isolate);
-
-    // Each individual compilation has its own dedicated [IsolateChannel], which
-    // closes once the compilation is finished.
-    var receivePort = ReceivePort();
-    isolate.sink.add((receivePort.sendPort, compilationId));
-
-    var channel = IsolateChannel<Uint8List>.connectReceive(receivePort);
-    channel.stream.listen(
-        (message) {
-          // The first byte of messages from isolates indicates whether the
-          // entire compilation is finished. Sending this as part of the message
-          // buffer rather than a separate message avoids a race condition where
-          // the host might send a new compilation request with the same ID as
-          // one that just finished before the [IsolateDispatcher] receives word
-          // that the isolate with that ID is done. See sass/dart-sass#2004.
-          if (message[0] == 1) {
-            channel.sink.close();
-            _activeIsolates.remove(compilationId);
-            _inactiveIsolates.add(isolate);
-            resource.release();
-          }
-          _channel.sink.add(Uint8List.sublistView(message, 1));
-        },
-        onError: (Object error, StackTrace stackTrace) =>
-            _handleError(error, stackTrace),
-        onDone: () {
-          if (_closed) isolate.sink.close();
-        });
-    _activeIsolates[compilationId] = channel.sink;
-    return channel.sink;
+    _sinks.add(channel.sink);
+
+    channel.stream.listen((message) {
+      // The first byte of messages from isolates indicates whether the
+      // entire compilation is finished. Sending this as part of the message
+      // buffer rather than a separate message avoids a race condition where
+      // the host might send a new compilation request with the same ID as
+      // one that just finished before the [IsolateDispatcher] receives word
+      // that the isolate with that ID is done. See sass/dart-sass#2004.
+      if (message[0] == 1) {
+        _mailboxes.remove(lease.id);
+        _inactive.add(lease);
+        lease.resource.release();
+      }
+      _channel.sink.add(Uint8List.sublistView(message, 1));
+    }, onError: (Object error, StackTrace stackTrace) {
+      _handleError(error, stackTrace);
+    }, onDone: () {
+      try {
+        mailbox.put(Uint8List(0));
```

Comment:
Why do we need to send an explicit "close" message when we've already received a close from the isolate?

Reply:
To reuse the isolate in the most efficient way, the logic is basically:

``` dart
while {
  mailbox.take();
  ...
} do (!_asyncError);
```

The mutex basically puts the thread to into sleep state until it receives a new message and wakes up. However, the mutex while waiting is in FFI calling native code, and therefore it is not managed by dart lifecycle until re-entering a Dart code frame back from native code. This means while the thread is sleeping, `isolate.kill()` has no effect until the thread wakes up and come back to dart code frame. Therefore, we need to send something to wake up the thread to make sure it will be gracefully terminated on next dart frame.

Reply:
> although it seems like sending the empty message in the same place as calling isolate.kill() might be clearer

You're right, that we can just send an empty message to break from the loop and the isolate will stop on its own.

Reply:
Actually I see that you have already refactored this part. And it looks good to me.